File: web-app/src/types/aggregator.ts
export interface LinkedProvider {
  id: number;
  provider_type: 'plaid' | 'akoya';
  institution_id?: string;
  institution_name?: string;
  status: 'active' | 'login_required' | 'error' | 'revoked';
  error_code?: string;
  error_message?: string;
  last_synced_at?: string;
  created_at: string;
  account_count: number;
}

export interface SyncResult {
  transactions_added: number;
  transactions_modified: number;
  transactions_removed: number;
  balances_updated: number;
  holdings_updated: number;
}

export interface LinkTokenResponse {
  link_token: string;
  expiration: string;
}

export interface AkoyaAuthUrlResponse {
  authorization_url: string;
  provider_id: string;
}

File: core-api/src/financial_analysis/api/schemas/aggregator.py
"""Pydantic schemas for bank data aggregation (Plaid + Akoya)."""

from datetime import datetime
from typing import Optional, List
from pydantic import BaseModel, Field, ConfigDict


class LinkTokenResponse(BaseModel):
    """Response from creating a Plaid Link token."""
    link_token: str
    expiration: str


class AkoyaAuthUrlResponse(BaseModel):
    """Response with Akoya OAuth authorization URL."""
    authorization_url: str
    provider_id: str


class PublicTokenExchange(BaseModel):
    """Request to exchange a Plaid public token."""
    public_token: str = Field(..., description="Plaid public token from Link")
    institution_id: Optional[str] = Field(None, description="Plaid institution ID")
    institution_name: Optional[str] = Field(None, description="Institution display name")


class AkoyaCallbackRequest(BaseModel):
    """Request to exchange an Akoya authorization code."""
    auth_code: str = Field(..., description="Authorization code from Akoya OAuth redirect")
    provider_id: str = Field(default="fidelity", description="Akoya provider ID")


class LinkedProviderResponse(BaseModel):
    """Response with linked provider details."""
    id: int
    provider_type: str
    institution_id: Optional[str]
    institution_name: Optional[str]
    status: str
    error_code: Optional[str]
    error_message: Optional[str]
    last_synced_at: Optional[datetime]
    created_at: datetime
    account_count: int = 0

    model_config = ConfigDict(from_attributes=True)


class SyncResultResponse(BaseModel):
    """Response from a sync operation."""
    transactions_added: int = 0
    transactions_modified: int = 0
    transactions_removed: int = 0
    balances_updated: int = 0
    holdings_updated: int = 0
